### ISSUE TITLE
2.6.10 Fill lookup table Clears the table not fills it

### ISSUE THREAD

**Issue by OverlanAI** [Program-O/Program-O#392]
In the SRAI admin section running (win10 wamp) and clicking Fill lookup table, We see it search as before but when it is done clears all entry's.
Over time it will add them in but only on demand during a conversation.
We cant tell if that is now WAI or an Ooops!

tnx

**Comment by OverlanAI** [comment]
Ahh seems i have to keep my mouse pointer over the pop-up that says wait and do not change to another page. (i did not change pages) just did not have the mouse pointer on the pop-up.
Still that doesn't seem right.

**Comment by Dave-Morton** [comment]
The way the "fill lookup table" function works is that it first clears out __all__ srai lookup entries, for _every_ bot, removes the search indexes for both the `aiml` and `srai_lookup` tables, adds a new index specifically intended to speed up the process (VERY different than the one used to search for potential AIML matches under normal operation), run said search to find AIML categories in the `template` that contain `<srai>` tags, generate a list of said occurrences, along with the ID of the AIML category containing those `<srai>` tags, store them in the emptied `srai_lookup` table, removes the index from the `aiml` table, and finally recreates the original indices for both the `srai_lookup` and `aiml` tables. The work with the indices during this process is implemented because not doing so messes up those original indices, and makes search speed for both of the tables painfully slow afterward. 
Now all of this takes time (roughly one second per 1,500 AIML categories, give or take a little), and the "popup" says specifically that you should __not__ leave the page till the process is complete. If you have one or more bots with a large number of AIML categories, even if none of them contain `<srai>` tags, it could take up to several minutes for this process to run its course.

**Comment by OverlanAI** [comment]
Yep, as long as i keep my mouse pointer over the spinning pop-up it will fill it again, any place else on that page just empties it and thats that.
I found that i had to keep the mouse pointer on the spinning pop-up until it is totally done, if i move my mouse pointer off of it then it just empties it..
Anyway no big deal, just had no idea i had to keep the pointer on an exact spot on that page was all.
tnx

**Comment by OverlanAI** [comment]
Removed 0 existing rows from the srai lookup table, and altered the indexes.Elapsed: 6.181 seconds.
Found 47,427 AIML categories that contain SRAI calls. Elapsed: 0.783 seconds.
Loaded 29,822 AIML categories that need to be added to the lookup table. Elapsed: 27.3 seconds.
Inserted 0 new entries into the SRAI lookup table. Elapsed: 7.82
Total elapsed time: 42.083 seconds.

None are shown.

**Comment by OverlanAI** [comment]
Would the size of the set cause it to end before completing?
I thought i had it figured out but when i changed to the Rue set (huge) it no longer fills the table.
My other set is basic ALICE set and it seems ok with that.
I have 2 instances of programO, using 2 different DBs both running current dev ect.
The ALICE one works and Rue does not and i have found no difference between the 2 instances other than the size of the sets.
Call me confused now.
tnx

**Comment by OverlanAI** [comment]
Well i did more test's (fresh install) by uploading a smaller AIML set (alice 1.9 w. minor edits) and find that the SRAI table fill's as expected.
Next, i dumped that set for my huge RUE set and uploaded it with no problems BUT it will not fill the SRAI table.
I can't see any errors or timeout (as seen above) just that it fails with 0 added after 42 sec..
I checked WAMP settings that could relate to some limits but they are well above anything it is doing.
I can only guess that ether it is hitting something in the process of filling it that it does not like and it bails out of the process OR simply was too big for it to deal with ( 47,427 AIML categories that contain SRAI calls).

Any suggestions?
tnx

**Comment by OverlanAI** [comment]
I changed from php 5.6x to 7.1x in wamp (easy to do btw) and ran the same test as above again.
It did not change the problem. I am stuck from going forward because with out the SRAI working the CPU usage on a an 8 core server can hit as high as 70% cpu usage on all cores during a normal chat.
When the SRAI is working (tables are filled) then it is all fine (drops to 5-10%) in heavy use at most.
Anyway i found no difference using php 7.x vers 5.6x.
tnx

**Comment by OverlanAI** [comment]
Changing from Mysql to MariaDB made no difference as i expected but sometimes you have to find out what a problem is not to find the problem.
Anyway i am stuck and cannot use the current dev. with my big aiml set.
So we have to hack around it using an older version for now.
tnx

**Comment by OverlanAI** [comment]
I am staring to wonder if it is bailing out when it hits an srai call that does not have a matching category.
It happens more than i like with no good way to find out what srai calls do not match in the first place.
So i am working on a tool that will try and test each srai call in the AIML database then output a list file of any that fail to find a working category for.
Why that may not be what is causing the main problem here, it still seems it would make a nice utility function for admin to use if they are running into Cat. not found problems.
Going further with the utility (i do not have my head around how yet) is to test for UN-reachable category's  within any aiml set.
But as far as it testing for non existent srai calls seems quite do able and we feel would help a lot. 
As near as i can track my problem down, is to the insert into section after it already knows what to fill the table with but fails if there are thousands of calls.
Why it bails out at that point, we have no idea. Just that it will fill it with small sets.
TNX (brb need a bigger hammer).

**Comment by OverlanAI** [comment]
Finally caught something in the logs when i goto update the srai and it fails to fill the table,
[error.zip](https://github.com/Program-O/Program-O/files/1891574/error.zip)
THE PDO log is what grows each time i try to upadte srai (ran it 2x to be sure that was what was adding it to the log.). The Normal error log may not have anything to do with it.
tnx

**Comment by OverlanAI** [comment]
Forgot the after shot doing the above workaround.

Removed 42,180 existing rows from the srai lookup table, and altered the indexes.Elapsed: 9.348 seconds.
Found 66,786 AIML categories that contain SRAI calls. Elapsed: 1.028 seconds.
Loaded 42,177 AIML categories that need to be added to the lookup table. Elapsed: 38.294 seconds.
Inserted 42,177 new entries into the SRAI lookup table. Elapsed: 8.276
Total elapsed time: 56.946 seconds.

Now doing near streaming chat tests.
tnx

**Comment by OverlanAI** [comment]
see issue 403, i think Hippo might be on to WTF and why.
But hes just 30 hahaha

**Comment by OverlanAI** [comment]
To tell you the truth i want this project to be 100% HELL yah so we can give it the attention it should have.
Problems are one thing, getting the world to help is hard.
I found a need then i found this and you got the good stuff.
tnx

**Comment by OverlanAI** [comment]
Im poking a few ass hats that did mit and some coders to take a look here. 
Not a joker in the deck but they all crap diamonds so who knows but they will look because they know me and my crew never makes noise about what we will do.
So get your tinfoil hats on ;-)
TNX (yah we hack them too).

**Comment by OverlanAI** [comment]
Anyway mort, they are coming and god help us!

**Comment by Dave-Morton** [comment]
@overlanAI you should shoot me an email.

**Comment by OverlanAI** [comment]
This is still an issue for me, It is hit and mis (sometimes it works sometimes not). I am still checking for possible server environment issues since i have run out of anything logical to look at.
Before the refactoring of php mysql this was not an issue.

TNX

**Comment by Dave-Morton** [comment]
I'm thinking that the current problem lies in the fact that the script is removing indices at the beginning, then replacing them once the script is done. this isn't necessary, really, and comes from a time when I didn't fully understand how indices work. The fix (of course) is to have all of the necessary indices put in place during the install process, and then just leave them alone. the only problem with making that change is that it would require a revamping of the install process, along with a fresh install being required after the update. This means a new Program O minor version (2.7.0), and up til now I haven't been willing to tackle that on my own. However, if it needs to happen it needs to happen, so I should just buckle down and get it done. But I want to fix some of the other bugs first, so this is down the ToDo list a little.

**Comment by OverlanAI** [comment]
Ahh, okay that explains a lot then. Index error are what tend to see more than any other error, usually something try's to index 0 but not always.
Anyway tnx for sharing that, that would at least explain the WHY.

**Comment by Dave-Morton** [comment]
I'd be interested in seeing what you come up with for indices for both the `aiml` and `srai_lookup` tables. While I've learned a lot over the past couple of years in this area, my "confidence" isn't as high as I would like it to be.

**Comment by OverlanAI** [comment]
Will do, i want it fixed the just as badly. I have 2 other possibility's to check where i do not see "*" at the top of the srai list and should be there because that srai is the Alpha and Omega no matter that the convo is about.  So could be something more simple or the result of index problems.
Anyway i need to check a few more things first b4 i can blame indexing error's but any or all of that would explain the bulk of the crazy stuff im running into.
TNX (i smell a rat)

**Comment by Dave-Morton** [comment]
the big problem with improving performance is the fact that the search for AIML categories that contain `<srai>` in the template is that the search will __always__ look something like this:

`select template, id from aiml where template like '%<srai>%';`

and that almost negates the power of an index. At least, this is what I've come to understand. I've tried to have phpMyAdmin `explain` the query, and the result more or less bears this out, as it shows NO advantage at all to adding an index for the `template` field.

**Comment by OverlanAI** [comment]
Yah have to, performance gets really bad Evan on an 8 core box with huge set and more than one bot using the older way.
At least i know what it is not so far but next i want to also look at the aiml index as that seems the same in both versions doing a quick look anyway.
So one may be correct and the other not and that would narrow it down just what part of the code the gremlin lives in.
After that ill get out my code hammer.

**Comment by OverlanAI** [comment]
Ahh, the template id's do not match up at all. In the working one the template ID and the aiml ID match correctly. That explains why the crazy crap happens in my #413 report.
And now i see why the SRAI ID is the way it is so that has nothing to do with it.
Still if the template ID does not match up with the correct  AIML id it makes and explains alot between the two.
So it could be you got it right but missed something with it matching up the template id to the correct aiml id it should be.
For example, In the working one i used the HELLO THERE template ID and it found HI HOW WAS YOUR DAY? .. that is correct.
In the non working one i found HELLO THERE and searched for the template ID in aiml and got back WHAT COLOR IS IT?
Not even close LOL.
So it may not be as bad as you may be thinking and more a simple OOPS!
Now to poke around in the code but getting closer now :-)

**Comment by OverlanAI** [comment]
that also would explain why * values ect are screwy as well. Not because of the AIML logic but because it is using the wrong template ID in the aiml that then changes the values to values they should not be depending on what the wrong template ID AIML does its trying to use.
So what i was calling random was not really but it totally would look that way.

**Comment by Dave-Morton** [comment]
You wouldn't need to do a data migration if a simple patch is performed to add the new column with a default of false. However, the SRAI lookup function would have to perform an "old school" search if there were no `true` entries in the new column of the `aiml` table. From that point on, the new indexed search should work as intended.

**Comment by OverlanAI** [comment]
It would not hurt to have a utility column added in any case Even if it is just for expanded uses should it really be needed. going forward but other wise ignored until then.
Anyway At least i know why i cant use current dev so i will stick with my patched (pre srai refactor 2.6.10) version for now. I am still making a generic aiml bot from the ALICE 1.9 (taking out all of the crap) so that will take me a few weeks anyway.
As for how to fix the current problem, i have to think about that a bit more but i see what you mean.
Im not so sure thats the best way to go about it yet.

**Comment by Dave-Morton** [comment]
Well, I've found _that_ problem, and the fix is fairly easy, but I've got so much code changes going on with my current `DEV` install of Program O it would be next to impossible to put what I'm working on aside to push a one line change in code, so I'll show you that line of code, instruct you on where to insert it, and let you test it to see if it helps. Here's that one line of code:

`$lookups = array_unique($lookups, SORT_REGULAR);`

Now currently, that line of code is on line #219 of my version, but I've made tons of changes to the file in a lot of places, so look for the following block of code:

```
    $now = microtime(true);
    $cet = round($now - $let, 3);
    $let = $now;
    $lookupsCount = count($lookups);
```
and place the new line of code directly above it. Then test it out and see if that fixes the problem. What I was seeing before I added that code was that there were specialized error log files being created (sometimes just one, sometimes hundreds) about one of two different things. Either they were complaining about "duplicate entries", or "array to string conversion". The duplicate entry log files only created from one to a small handful, but the array to string conversion created literally hundreds of log files. Either way, by adding that line of code, ALL of those log files were no longer created.

**Comment by OverlanAI** [comment]
Lol, what file name please?

**Comment by OverlanAI** [comment]
Removed 28,153 existing rows from the srai lookup table, and altered the indexes.Elapsed: 6.774 seconds.
Found 37,057 AIML categories that contain SRAI calls. Elapsed: 0.535 seconds.
Loaded 6,524 AIML categories that need to be added to the lookup table. Elapsed: 18.864 seconds.
Inserted 6,524 new entries into the SRAI lookup table. Elapsed: 5.421
Total elapsed time: 31.594 seconds.

Looking good sofar, That is about the correct number of SRAI for my set.
Ill test the change for a few days and let you know but at least the numbers look correct now.

**Comment by Dave-Morton** [comment]
If you look through that block of code (about 50 lines, I think, in either direction) and comment out all lines that call `db_write()` (EXCEPT for the one that deletes all current entries), you'll cut that total elapsed time in half. Just sayin'. 😉 

I'm completely rewriting that entire function so that it won't mess with the indexes at all. Additionally, I fully expect that by searching the new `has_srai` field for values of _true_, the searching/loading should also take considerably less time, but that still remains to be seen.

**Comment by OverlanAI** [comment]
Lol took it a whole hot 31 seconds so big woop.
More worried about the logic being correct than performance at this point and if that takes care of it can come back to performance tweaks later.
Anyway im setting my middle-ware to put that in meat grinder mode for a few days and see.

**Comment by OverlanAI** [comment]
Lovely, SRAI IDs to AIML  now match up correctly and also i reloaded a fresh bot to make sure it was not choking like it was on a fresh new bot.
+1 on this sofar, now to run all of the other tests we have a few K rounds of streaming chat.

**Comment by OverlanAI** [comment]
Well slight problem, the srai IDs match up ok so thats working like it should but the srai pattern is not sorted so that * is always the first srai pattern found when or if it cant find any match.
So when it can not find something it should use * so there is always an alpha and omega to fall thru to.
This is caused when an srai call is made but there is no match for it.
What it does now, it uses the 1st thing it finds at the top of srai list so that tends to mess things up if there is a mistake in the AIML. In the older version * was always at the top of the srai list so it would fail to a pickup line and continue in a more sensible way.
Anyway seems the SRAI patterns are not sorted correctly or at all but that is a guess on my part.
as long as the first thing it finds is star then that should end the srai issues that i know of.
Thats about as good of a description as i can think to tell you.

**Comment by OverlanAI** [comment]
SRAI Search ID numbers grow by the same amount of total SRAI found even if nothing has changed from the last time the table was cleared.
That may be nothing but im hitting SRAI Seach ID numbers at 1931513 and the next time i ran it (no changes) it jumped to 1938273, when i subtract one from the other i have the exact number of SRAI calls.
Im not sure why it would be that high with so few entries in the table.
Anyway mostly just making a note about it since that number is getting really high.
If anything it should have some check if nothing has changed in order to reduce the retching up like that.
Ill guess that is really just the value mysql auto adds.
Mostly this is just a foot now for now.

**Comment by OverlanAI** [comment]
YAY me.. added in TRUNCATE SRAI_LOOKUP fist then Fill it.
Now the SRAI ID starts at 1 again.
So someplace it is not doing that type of purge to the table that results in the above, since i do not want to have to load up php Myadmin each time i just hacked that in as TRUNCATE for now.
This is all still just foot notes since at some point it will all be re-done anyway.

**Comment by Dave-Morton** [comment]
That' fine if you only have one chatbot, or if you want to do the SRAI lookups for all your chatbots at once, but if you have multiple chatbots, and only want to do lookups for that one bot, then truncating the table is a bad idea. Just something to consider.

**Comment by OverlanAI** [comment]
Good point, i do have more than one bot but it has a parent so i did not notice any ill effect but i see what you mean.
Since i found other issues with multi-bots anyway ill just concentrate on the primary bot for now.
If nothing else i just confirmed that the ID had nothing to do with the other oddness at least.
From what your saying is sounds like doing the same thing for the AIML (then reload the sets) would also not change much for a single bot since the AIML IDs are also 7 digits long like the srai IDs were.
Anyway i wanted to see what if any effect it would have with some of the existing issues but as i suspected, had nothing to do it.
For now ill let it truncate just to keep things in check why i debug further. It takes no huge amount of time to clean start anyway. At least i know what the problem is not lol.

**Comment by OverlanAI** [comment]
Im a bit fuzzy about parent bot when it comes to performance Vrs A second bot that is on its own.
With a parent bot bot 2 has no real impact on performance that is that bad (depending on brain structure).

However if each bot is on its own (one has nothing to do with the other) then performance goes right in the crapper. 

I looked a some of your code a bit i wonder why in the case of non-parent bot's that each cant just have there own SRAI table.

Perhaps im not quite understanding the difference but seems if independent bots each have there own tables, performance would be a non-issue. 
Just thinking it could avoid many bugs if each bot has Independence tables and only would need to share IF the user wants to make use of a parent or layers of them depending on the brain structure there making.

Anyway i may be thinking about this all wrong.

tnx

**Comment by Dave-Morton** [comment]
I'm still stuck in the hospital, so don't really have the means to get into this right now. Let me get home, and get back into something resembling "normal", and I'll be happy to go over this with you then, ok?

**Comment by OverlanAI** [comment]
np man, clap shots are hell :)
we are thinking of c# as a side car to this project as mysql (screw php) as helpers as part of progamO.
means mixed code but dam.. PhP cant but it can use the hell yah of C#.
Take all the crap php cant do well and call on C# via php ..
Why the hell not?
Who says all this has to be just PHP anyway?
Your project your crew, but dam think about php calls to c# that deals with mysql crap... mm yummy!
when you do that there is no slow :)
mean time get well soon, hope the nurse is sexy .. mmm

**Comment by OverlanAI** [comment]
in the sim world we mix coding to get the speed, called mods the core sends command to. In one client we make it uses all languages as code to make one thing. 
Not just php not just C# name it.. 
When it compiles .. man thats the good stuff!

**Comment by OverlanAI** [comment]
just saying theres other way to skin this cat and dam this will be as good as there is

**Comment by OverlanAI** [comment]
btw man, the resion why the coders i know with the gold hammers do not chime in here is not because they dont want to.
They are bound as coders to other projects via there corp. and risk lossing there cred and jobs by doing side projects.
Its not that they cant or do not care but they can not only loose there jobs but also as trusted coders.
So it is not so easy as users think and never for the resions they they think.
the gold hammers get payed big but gave there soul to the corp.
Hackers like me and crew dont give a crap and will do it anyway.
That sends a message to the gold hammers and makes your project center to whats next!
i looked you up and they did too . 
All said mort is dammed good :)
Im old man old bread board one wire at a time to make a computer type..
But no matter how good the chips are, no matter how good the binary probe is..
it takes minds like you to make it matter.

**Comment by OverlanAI** [comment]
Anyway let the nurses do there jobs and get well soon!

**Comment by OverlanAI** [comment]
i sent all this crap off to the gold hammers and the best crew i know to take a look. 
I want learn AI i want perform solved and it is all given back to you.
What happens next i cant say but it wil all be given back to you with cred and fist say.
thats the best i can do.
ProjectO is the best and there will be no doubts who the master is.
besides the gold hammers can get sewed by law of they umm lol
You got the good stuff man and they know me so.
Get well soon man.
tnx (love real coders).

**Comment by OverlanAI** [comment]
Also my girl that knows PHP alot pokes me and says your one of the best and asks what food do yu like to eat.. odd huh?

**Comment by OverlanAI** [comment]
im C# she is PHP lol but it works.. anyway she says do what the nurses say (guessing your grumpy like me) and get out of there fast.
(hospitals are not place for sick coders).
tnx
https://www.youtube.com/watch?v=vpCTMddpQNs
mmm sexy :)

**Comment by OverlanAI** [comment]
and to my crew its more like this
https://www.youtube.com/watch?v=qpMvS1Q1sos
cuz we got that right mmm.
so get your ass fixed .. best of the best cant be you!
smile

**Comment by OverlanAI** [comment]
take your pick but you got this right !

**Comment by OverlanAI** [comment]
mean time im breaking down the php into sections that choke and telling them to show me there best.
 when there done telling 100 reasions why i cant do a thing, the cool kids will show there wrong.
ie" cant just got done!".
Crew dont do it like that, cant be done is why we do it.
smile man
https://www.youtube.com/watch?v=3f2g4RMfhS0
get well soon and pull up up yor pants  man.
Opensim need you man.
Real AI and thats why we care!

**Comment by OverlanAI** [comment]
https://www.youtube.com/watch?v=3f2g4RMfhS0
sing to the doc and get home soon

**Comment by OverlanAI** [comment]
and yea im a prick but you got the good stuf! mmm

**Comment by OverlanAI** [comment]
btw i looked you up .. i have to ask do you know c# ?

**Comment by OverlanAI** [comment]
mean time why they fix you
https://www.youtube.com/watch?v=Ypkv0HeUvTc
hacker thing lol

**Comment by OverlanAI** [comment]
get your ass up and getr back here soon man. we need yah. btw do you have isues with live stream ?

**Comment by OverlanAI** [comment]
i got that bandwidth to stream you live is you got the time
it will be about opensim gets projectO  or something but real time
Like i said .. god help yah.. mmm
we only play rock here.
https://www.youtube.com/watch?v=07Y0cy-nvAg
get it right :

**Comment by OverlanAI** [comment]
smile

**Comment by OverlanAI** [comment]
anyway mort , there going to look now becuse the pricks owe me and my crew.
I cant say what happens with real coders that cant  agree that the sun is red but at lest i did my part,
For my own hacks.. i code to show off, cuz cant just got done.
Thats why i put you and the best minds to it.
I love "Cant be done!" just got done .. mmm sexy to me and crew
Mean time get out of the med lab man,
We need yah!

**Comment by OverlanAI** [comment]
after they are done with you in that place..
play this
https://www.youtube.com/watch?v=VYIxUM-CnSw
get well soon

**Comment by Dave-Morton** [comment]
Ok, a couple of things here, just to be clear:

1. this is a family-friendly environment, so certain comments are not appropriate. I've edited out the ones I saw, and will continue to do so in the future.

2. Migrating to another language, such as C#, is NOT on the roadmap, and discussions of doing so, even to speculate, are just a waste of effort and time here. If you want to do so outside this project, you have the collective blessings and well wishes of everyone on the team, so knock yourself out. Just don't bother discussing it here.

3. The Issues section here on GitHub is intended to be used to directly discuss current problems, bugs and difficulties within the project. Discussions outside those topic parameters dilute our ability to make fixes and improvements, so it's best to stick strictly to the subject matter at hand, and not wander off topic.

4. Making multiple posts with little to no additional information that's relevant to the current issue also dilutes our efforts. In the future, it's appreciated if you limit your posts to the fewest possible, so as not to "spam" the issue thread.

The bottom line here is that you're getting a bit out of hand, and you need to curb yourself. I truly appreciate the contribution you've been making to the project, but your actions are actually making it harder for us to make any positive change. Please let's keep a civil tongue, stay on topic, and try not to flood the threads here with bunches of rapid fire posts. this need not get ugly if you see this as a request to help, instead of an order to grow up.

**Comment by OverlanAI** [comment]
your W-hat pissed off the wrong hackers to day. Take his keys away,, he is useless
